torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,143 @@
+                )
+                for i in range(n_dim)
+            }
+        )
+
+    @property
+    def seq_len(self) -> int:
+        """Dimension of flattened data, e.g., time * height * width"""
+        return np.prod(self.shape)
+
+    @property
+    def decode_idxs(self):
+        """Indices along the dims of data, e.g., ``(time, height, width)``."""
+        return list(itertools.product(*[range(s) for s in self.shape]))
+
+    def _broadcast(self, emb: torch.Tensor, i: int) -> torch.Tensor:
+        """Broadcasts the ``i``-th embedding matrix ``(self.shape[i], self.embedding_dim // n_dim)`` along the other
+        dims of ``self.shape``. The embedding dim is kept at ``self.dim``.
+
+        For example::
+
+            >>> pos_emb = BroadcastPositionEmbedding(shape=(2, 4), embedding_dim=6)
+            >>> print(pos_emb.embedding["d_0"]).shape
+            torch.Size([2, 3])
+            >>> out = pos_emb._broadcast(torch.tensor([[0, 0, 0],[0, 0, 1]]), i=0)
+            >>> print(out)
+            tensor([[[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
+                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]]])
+
+        The input is broadcasted along dim ``4`` since it's the ``0``-th embedding constructed w.r.t dim ``2``.
+        """
+        if self.dim == -1:
+            # (1, 1, ..., 1, self.shape[i], 1, ..., -1)
+            emb = emb.view(
+                1, *((1,) * i), self.shape[i], *((1,) * (self.n_dim - i - 1)), -1
+            )
+            # (1, *self.shape, -1)
+            emb = emb.expand(1, *self.shape, -1)
+        else:
+            emb = emb.view(
+                1, -1, *((1,) * i), self.shape[i], *((1,) * (self.n_dim - i - 1))
+            )
+            emb = emb.expand(1, -1, *self.shape)
+
+        return emb
+
+    def _decode(
+        self, decode_step: int, embeddings: torch.Tensor, x_shape: torch.Size
+    ) -> torch.Tensor:
+        """Returns the embedding vector immediately before the decoding location."""
+        decode_idx = self.decode_idxs[decode_step - 1]
+        if self.dim == -1:
+            embeddings = tensor_slice(
+                embeddings,
+                [0, *decode_idx, 0],
+                [x_shape[0], *(1,) * self.n_dim, x_shape[-1]],
+            )
+        elif self.dim == 1:
+            embeddings = tensor_slice(
+                embeddings,
+                [0, 0, *decode_idx],
+                [x_shape[0], x_shape[1], *(1,) * self.n_dim],
+            )
+
+        return embeddings
+
+    def forward(
+        self, x: Optional[torch.Tensor] = None, decode_step: Optional[int] = None
+    ) -> torch.Tensor:
+        embeddings = []
+        for i in range(self.n_dim):
+            emb = self.embedding[f"d_{i}"]
+            emb = self._broadcast(emb, i)
+            embeddings.append(emb)
+
+        embeddings = torch.cat(embeddings, dim=self.dim)
+
+        if decode_step is not None:
+            embeddings = self._decode(decode_step, embeddings, x.shape)
```

Comment:
why not take in the shape of x instead of x?

Reply:
`_decode` only requires `x_shape` but `x` is passed in as a tensor at the call site so the signature of `forward` needs to remain as `x`. Does that answer your question?